ksonnet-util/kausal.libsonnet

```diff
@@ -143,29 +143,25 @@ k {
 
     deployment+: {
       new(name, replicas, containers, podLabels={})::
-        super.new(name, replicas, containers, podLabels { name: name }) +
+        local labels = podLabels {name: name};
+        super.new(name, replicas, containers, labels) +
 
         // We want to specify a minReadySeconds on every deployment, so we get some
         // very basic canarying, for instance, with bad arguments.
         super.mixin.spec.withMinReadySeconds(10) +
 
         // We want to add a sensible default for the number of old deployments
         // handing around.
-        super.mixin.spec.withRevisionHistoryLimit(10),
+        super.mixin.spec.withRevisionHistoryLimit(10) +
+
+        // required matchLabels selector
+        super.mixin.spec.selector.withMatchLabels(labels),
     },
 
     statefulSet+: {
       new(name, replicas, containers, volumeClaims, podLabels={})::
-        { kind: 'StatefulSet'} + 
-        { apiVersion: 'apps/v1beta1' } + 
-        self.mixin.metadata.withName(name) + 
-        self.mixin.spec.withReplicas(replicas) + 
-        self.mixin.spec.template.spec.withContainers(containers) + 
-        self.mixin.spec.template.metadata.withLabels(podLabels { name: name }) +
-        (if std.length(volumeClaims)>0
-        then self.mixin.spec.withVolumeClaimTemplates(volumeClaims)
-        else {}) +
```

Comment:
you've just removed some important code here.

If we have a statefulset that has no volumes (for some apps, we use a statefulset for the consistent pod naming, not its volume persistence), then by default `k.libsonnet` will generate a resource that will **always** generate a diff (IIRC it generates an empty list, which Kubernetes drops, thus Jsonnet and K8s are always diferent). These three lines (which necessitate this whole block) cause Jsonnet to render a statefulset that does NOT include the volumeclaim block if no volumes are used.

Reply:
Fixed in 332f7eb4e62aabe93caeb53e0a755d4cd69dd356